File: extract-porter-app.mjs
/**
 * One-time / on-demand: extract the inline Babel bundle from indexTEST.html → vite-scaffold/src/porter-app.jsx
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const HTML = path.join(ROOT, 'indexTEST.html');
const OUT = path.join(ROOT, 'vite-scaffold', 'src', 'porter-app.jsx');

const s = fs.readFileSync(HTML, 'utf8');
const m = s.match(/<script type="text\/babel">\s*\n/);
if (!m || m.index === undefined) throw new Error('Could not find inline babel script');
const afterOpen = m.index + m[0].length;
const rootIdx = s.indexOf('const root = ReactDOM.createRoot', afterOpen);
if (rootIdx < 0) throw new Error('Could not find createRoot block');
let body = s.slice(afterOpen, rootIdx).trimEnd();
// Remove destructuring from global React (we import hooks from 'react')
body = body.replace(
  /^\s*const\s*\{\s*useState\s*,\s*useMemo\s*,\s*useEffect\s*,\s*useRef\s*\}\s*=\s*React\s*;\s*\n/m,
  '',
);
const preamble = `import React, { useState, useMemo, useEffect, useRef } from 'react';

`;
const postamble = `

export { App, ErrorBoundary };
`;
fs.mkdirSync(path.dirname(OUT), { recursive: true });
fs.writeFileSync(OUT, preamble + body + postamble, 'utf8');
console.log('Wrote', OUT, `(${(preamble + body + postamble).length} chars)`);

File: extract_porter_app.py
"""Extract inline Babel bundle from indexTEST.html → vite-scaffold/src/porter-app.jsx."""
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
html = (ROOT / "indexTEST.html").read_text(encoding="utf-8")
# Inline bundle only (not <script type="text/babel" src="...">)
m = re.search(r"<script type=\"text/babel\">\s*\n", html)
if not m:
    raise SystemExit("missing inline babel script")
after_open = m.end()
root_idx = html.find("const root = ReactDOM.createRoot", after_open)
if root_idx < 0:
    raise SystemExit("missing createRoot")
body = html[after_open:root_idx].rstrip()
body = re.sub(
    r"^\s*const\s*\{\s*useState\s*,\s*useMemo\s*,\s*useEffect\s*,\s*useRef\s*\}\s*=\s*React\s*;\s*\n",
    "",
    body,
    count=1,
    flags=re.M,
)
out_dir = ROOT / "vite-scaffold" / "src"
out_dir.mkdir(parents=True, exist_ok=True)
preamble = "import React, { useState, useMemo, useEffect, useRef } from 'react';\n\n"
postamble = "\n\nexport { App, ErrorBoundary };\n"
out = Path(sys.argv[1]) if len(sys.argv) > 1 else (out_dir / "porter-app.jsx")
out.parent.mkdir(parents=True, exist_ok=True)
out.write_text(preamble + body + postamble, encoding="utf-8")
print("Wrote", out, len(preamble + body + postamble), "chars")
